IA-3762 useLoadedData onError callback

msilva-broad opened the pull request:
- part of the work for Export Analysis Modal TS/Unit-Tests
- concurrent work by other TS development work surfaced the utility of adding this callback.
- optional onError callback arg now available at useLoadedData hook creation time.
- useful for adding side-effects like ReportError(...) while still having the consuming hook emit LoadedState<S>

msilva-broad committed:
IA-3762 useLoadedData onError callback
- part of the work for Export Analysis Modal TS/Unit-Tests
- optional onError callback arg now available at useLoadedData hook creation time.
- useful for adding side-effects like ReportError(...) while still having the consuming hook emit LoadedState<S>

s-rubenstein commented:
Swapped in @pshapiro4broad for the Data Explorer team, as he is our current sprint goalie.

pshapiro4broad commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -50,11 +64,15 @@ export const useLoadedData = <T>(): UseLoadedDataResult<T> => {
-      })
+      }
+      setLoadedData(errorResult)
+      if (args.onError) {
+        args.onError(errorResult)
+      }
```
Can this be written as
```suggestion
      args?.onError(errorResult)
```

msilva-broad replied:
it could (sort of)... but I find folks get confused and/or over/misuse this operator... it only validates the LEFT item in the chain (args) in this case to not be null or undefined)... in this case args is already guaranteed to not be null because the type safety on line 48 above this where args is declared.  What I think you are suggesting is this:
`args.onError?.(errorResult)`
as detailed in the "optional call"  paragraph of https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining

I can give this a try... as long as the instrumentation of proper unit test branch coverage check out, I'll get it in before merge.

msilva-broad replied:
uh oh.... tried:
`args.onError?.(errorResult)`
and the branch coverage falsely reported 100% without me testing the args.onError == null or undefined case.

Looks like I will need to add a note to advise caution/should-avoid on this since it currently cripples code coverage checks for important cases that should be tested.

pshapiro4broad replied:
Yes, `args.onError?.(errorResult)` is what I meant to type

pshapiro4broad replied:
> and the branch coverage falsely reported 100% without me testing the args.onError == null or undefined case.

IMO this is fine and I wouldn't expect to have to write a test to validate the `?.` is working properly. That's the benefit of using a language builtin vs a `? :` or `||` operation to handle undefined state.

msilva-broad replied:
the danger is that these are syntactic sugar for code equivalents that are code branches.   In this example, not being alerted of the branch coverage gap means that there is a good chance that a test with good assertions around the onError callback correctly being called if provided never gets written.  Code changes could break this expectation and there would be no signal that consumers' expectations just got broken.

pshapiro4broad replied:
I understand that typescript converts its code to javascript so its "primitives" are really implemented as javascript code. And in this case, yes, `x?.y` is converted to a conditional (I think it becomes `x && x.y`).

What I'm saying is that, unlike `x && x.y`, there's no need to test the case where `x` is false, because we know that typescript generates the correct code. As long as the positive case has a test, that should be sufficient. If you're saying that the coverage report doesn't require testing the _positive_ case, then I agree, it's a bug.

msilva-broad replied:
right... and that is the danger of some of these convenience patterns... they easily hide code branches, in sometimes significantly negative ways.  Let's define code branches as "the code can produce more then one outcome", whether thru more conventional if statements, language conveniences, or anything in between. The underlying mechanic doesn't matter... just that it produces more then one possible outcome in the code it's being used in.   In the example above, as long as one test hits the line args.onError.?(errorResult) with the falsy case, it's considered completely covered... hence creating the positive case not needing to be tested scenario.

The Typescript to Javascript flow is it's own beast... First you have the conversion to be careful of (usually babel.js under the hood), and then there is the intrinsically dynamic elements of JavaScript itself.  I am curious to look into possible tooling solutions out there to avoid the branch hiding of some of these... but that's another track.

pshapiro4broad replied:
I ran some tests and it looks like the jest coverage data does take both branches into consideration when generating branch coverage data for `x?.y()`. Here's what I did. The first run doesn't use `onError`, the second does, and the final run executes all loaded data tests. Note that the branch coverage is the second column.
```
$ yarn run react-scripts test --coverage --testNamePattern='^useLoadedData hook handles error state from Fetch Response as error$' --runTestsByPath src/libs/ajax/loaded-data/useLoadedData.test.ts | grep useLoadedData.ts
useLoadedData.ts                                 |    92.3 |       60 |     100 |    92.3 | 61
$ yarn run react-scripts test --coverage --testNamePattern='^useLoadedData hook handles ready state, then later error state$' --runTestsByPath src/libs/ajax/loaded-data/useLoadedData.test.ts | grep useLoadedData.ts
useLoadedData.ts                                 |     100 |       60 |     100 |     100 | 50,66
$ yarn run react-scripts test --coverage --testNamePattern='^useLoadedData hook'  --runTestsByPath src/libs/ajax/loaded-data/useLoadedData.test.ts | grep useLoadedData.ts
useLoadedData.ts                                 |     100 |      100 |     100 |     100 |
```
Based on this, it looks like there's no harm in using `?.` with respect to code coverage.

msilva-broad replied:
created a UIE ticket to track the exploration of a tooling fix:
https://broadworkbench.atlassian.net/browse/UIE-68

pshapiro4broad commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -31,7 +44,8 @@ export type UseLoadedDataResult<T> = [
  */
-export const useLoadedData = <T>(): UseLoadedDataResult<T> => {
+export const useLoadedData = <T>(hookArgs?: UseLoadedDataArgs<T>): UseLoadedDataResult<T> => {
+  const args: UseLoadedDataArgs<T> = hookArgs ? hookArgs : {}
```
I think `||` is more common in terra-ui for this concept
```suggestion
  const args: UseLoadedDataArgs<T> = hookArgs || {}
```

msilva-broad replied:
i like it.  Let me get this in for final commit.

msilva-broad replied:
scratch that... looks like this would cripple unit test branch coverage and falsely report 100% branch coverage even if the true and false paths are not both being tested.  I will add this as caution/avoid to the code guidance...

pshapiro4broad replied:
Could it instead be replaced with a `?.` later on in the code?

msilva-broad replied:
removed by switching over to default function arg value.

pshapiro4broad commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -31,7 +44,8 @@ export type UseLoadedDataResult<T> = [
  */
-export const useLoadedData = <T>(): UseLoadedDataResult<T> => {
+export const useLoadedData = <T>(hookArgs?: UseLoadedDataArgs<T>): UseLoadedDataResult<T> => {
+  const args: UseLoadedDataArgs<T> = hookArgs ? hookArgs : {}
```
Is it not possible for the type to be inferred here?
```suggestion
  const arg = hookArgs ? hookArgs : {}
```

msilva-broad replied:
this is for added type safety so that any future code changes don't change the attended outcome of a usable set of args according to the type provided.  Think of scenarios where a required arg is added to the type later... then the fallback of empty object {} would be a wanted type error here.

pshapiro4broad replied:
Could a default parameter value be used here? If so, it seems like that would fix a few problems.
```
export const useLoadedData = <T>(args: UseLoadedDataArgs<T> = {}): UseLoadedDataResult<T> => {
```

msilva-broad replied:
good call!  Let me try that...

msilva-broad replied:
done

pshapiro4broad commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -50,11 +64,15 @@ export const useLoadedData = <T>(): UseLoadedDataResult<T> => {
-      })
+      }
+      setLoadedData(errorResult)
+      if (args.onError) {
```
If `args` isn't used until this part of the code, can its definition be moved here too? Or is there some performance reason to only declare it at the top. I find it easier to read code if an object's definition is as close to its use as possible.

msilva-broad replied:
my instinct is to handle default/fallback behavior for arguments at the top of a code block so that the normalizing happens first-thing.  While I agree with the core principle of define-near-use... argument normalization is something I've nearly always seen at the top of function blocks in JS/TS

pshapiro4broad commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -50,11 +64,15 @@ export const useLoadedData = <T>(): UseLoadedDataResult<T> => {
         Error(await err.text()) :
         err
-      setLoadedData({
+      const errorResult: ErrorState<T, unknown> = {
```
It's a shame that the type can't be inferred here. In Java there's an idiom of static factories that can be used to infer a generic's type arguments from the factory parameters which avoids this kind of redundancy. But I don't know if there's an idiomatic way to do this in typescript.

msilva-broad replied:
I think we want to be careful with "over-inferring" in Typescript.  I greatly value clarity of developer intent over "brevity".  The starting point here is JavaScript, a full-on dynamic language, as opposed to Java that is statically typed from the get-go.  We want to bias on the side of leveraging Typescript to express developer intent and adding an extra layer of type safety for later code edits.

pshapiro4broad commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -50,11 +64,15 @@ export const useLoadedData = <T>(): UseLoadedDataResult<T> => {
         error
-      })
+      }
+      setLoadedData(errorResult)
```
It seems odd to me that the error result is now put in two different places, so client code has two different ways to be notified when an error occurs. Is there a need for both code paths? I don't see any use of this code path in this PR, and the linked ticket (https://broadworkbench.atlassian.net/browse/IA-3762) doesn't mention anything about adding this API.

My usual preference is to only add code when it's needed, to avoid keeping around unused code.

msilva-broad replied:
the need came from nearby developer activity during the IA-3762 work... as the example states, onError is to accommodate patterns for side-effects.  A general desire is to bias towards allowing side-effects to be handled in layers outside the visual component... which this allows for by letting a consuming hook of the useInterestingData or useInterestingState pattern handle this side-effect and let the visual component just respond to the at-rest state object of { status: "Error", ... }

msilva-broad replied:
I updated the PR comments to highlight the source of the need... and will make sure it gets into the merged commit description.

pshapiro4broad replied:
OK, although I'd prefer to see a feature added only as its use is added. Otherwise it's hard to tell what code is being used or not in a code base. If I were to PR onto the code as-is, I might remove this feature, since it has no uses.

pshapiro4broad reviewed (approved):
The code looks OK although I have a few comments and questions.

My main question is why is this code being added, if it's not used in the code base. I think it's better to only add code when it's needed, and remove code when it's no longer used.

jdcanas reviewed (approved):
Most of the discussion points have already been noted. Will give this a thumb pending the resolution of those

msilva-broad committed:
IA-3762 useLoadedData onError callback
- PR feedback
- switched to default arg value in hook signature
- also added usage detail to onError jsdoc comment

pshapiro4broad commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -1,7 +1,23 @@
+   *   onError: (errState) => ReportError(errState.error)
+   * })
+   */
+  onError?: (state: ErrorState<T, unknown>) => void
```
One last idea, it occurs to me that if the default parameter list had a no-op operation for `onError` then you no longer need to make it optional and wouldn't need to check for it below. e.g.
```
  onError: (state: ErrorState<T, unknown>) => void

[...]

export const useLoadedData = <T>(hookArgs: UseLoadedDataArgs<T> = { onError: () => {} }): UseLoadedDataResult<T> => {

[...]

      hookArgs.onError(errorResult)

```
Unless there's a case where you'd want to allow the caller to explicitly pass it in as `undefined`.

msilva-broad replied:
hmmm...  a good thought, but feels a little heavy-handed just to avoid a bit of internal handling.

The spirit of going with a single args object as opposed to a growing list of function arguments is for a number of longterm benefits:
- argument "names" are auto-documented since they must be given as part of the arg object by the consumer, making code more readable.
- argument order no longer matters
- argument optional/required no longer conflicts with argument list order
- a number of object conveniences, including TS.
- contract and consumer refactoring tends to be less painful later.

And long-term what I tend to do with larger argument objects is to have a default object with the default if any, and merge the incoming arg object over the default one at the top of the function/class.  But here I am keeping things simple since we may never need more than this one arg.

pshapiro4broad replied:
This approach is still using a single args object. The only difference is that the single args object default value isn't empty anymore. So all the other attributes you mentioned are still true.

msilva-broad committed:
Merge branch 'dev' into IA-3762_useLoadedData-onError
